### PR TITLE
fix: fix loading indicator to support ghost loader

### PR DESCRIPTION
**Pull request by chintankavathia**
**What kind of change does this PR introduce?** (check one with "x")

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)
Today, datatable with pagination on server side doesn't have ghost loaders to indicate data is being loaded.

**What is the new behavior?**
This PR adds option to enable ghost loader which can be used to indicate data is being loaded from server while doing pagination

**Does this PR introduce a breaking change?** (check one with "x")

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

https://user-images.githubusercontent.com/3906709/153598878-93bf579e-2dc0-4389-bfae-9389cd22e762.mov

### CODE REVIEW

**Comment by timowolf** [comment]
Many thanks @chintankavathia. I will review as quick as possible.

**Comment by chintankavathia** [comment]
> Many thanks @chintankavathia. I will review as quick as possible.

Thanks @timowolf . Meanwhile Is there any other item which I can look into ?

File: projects/ngx-datatable/src/lib/components/body/body.component.ts [hunk 1]
```diff
@@ -135,6 +138,7 @@ export class DataTableBodyComponent implements OnInit, OnDestroy {
   @Input() scrollbarV: boolean;
   @Input() scrollbarH: boolean;
   @Input() loadingIndicator: boolean;
+  @Input() ghostLoader: boolean;
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/components/body/body.component.ts:141]: Thinking about the naming here: would it make sense to align the naming closer to the existing `loadingIndicator` one, e.g:

```suggestion
  @Input() ghostLoadingIndicator: boolean;
```

On a second note, it seems that the term `skeleton` load is more widely used as opposed to `ghost` loader. Would it make sense to all align it to `skeleton`, e.g:

```suggestion
  @Input() skeletonLoadingIndicator: boolean;
```

@chintankavathia @timowolf WDYT?

File: projects/ngx-datatable/src/lib/components/body/ghost-loader/ghost-loader.component.html [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<div class="ghost-loader">
+    <div class="ghost-element" *ngFor="let item of [].constructor(pageSize)">
+        <div class="line" [ngStyle]="{'width': col?.width + 'px'}" *ngFor="let col of columns"></div>
+    </div>
+</div>
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/components/body/ghost-loader/ghost-loader.component.html:5]: - The library exclusively uses 2-space indents, we hence should keep it the same.
- Since you only bind a single style, we can use a direct style binding.

```suggestion
<div class="ghost-loader">
  <div class="ghost-element" *ngFor="let item of [].constructor(pageSize)">
    <div class="line" [style.width]="col?.width + 'px'" *ngFor="let col of columns"></div>
  </div>
</div>
```

File: projects/ngx-datatable/src/lib/components/datatable.component.scss [hunk 3]
```diff
@@ -216,7 +216,7 @@
     position: relative;
     z-index: 10;
     display: block;
-
+    overflow: hidden;
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/components/datatable.component.scss:219]: To remain consistent with the rest of the spacing and keep the semantics aligned:

```suggestion
    overflow: hidden;

```

File: projects/ngx-datatable/src/lib/themes/material.scss [hunk 4]
```diff
@@ -466,6 +471,34 @@ $datatable-summary-row-background-hover: #ddd !default;
...
+    background: linear-gradient(90deg, $datatable-ghost-loader-background-light, $datatable-ghost-loader-background, $datatable-ghost-loader-background-light) 0 0/ 80vh 100% fixed;
+    background-color: $datatable-ghost-loader-background-dark;
+    animation: ghost 750ms infinite linear;
+    
+  }
+
+  .ghost-element {
+      display: flex;
+  }
+}
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/themes/material.scss:500]: - Might make sense to add a chapter comment here.
- Slightly more consistent indents and spacing.

```suggestion
/**
 * Ghost loader animation
 */
@keyframes ghost {
  from {
    background-position: 0vw 0;
  }
  to {
    background-position: 100vw 0;
  }
}

.ghost-loader {
  position: sticky;
  top: 20px;
  height: 0;
  background-color: $ngx-datatable-default-background;

  .line {
    width: 100%;
    height: 12px;
    margin: 0.9rem 1.2rem;
    background: linear-gradient(90deg, $datatable-ghost-loader-background-light, $datatable-ghost-loader-background, $datatable-ghost-loader-background-light) 0 0/ 80vh 100% fixed;
    background-color: $datatable-ghost-loader-background-dark;
    animation: ghost 750ms infinite linear;
  }

  .ghost-element {
    display: flex;
  }
}
```

File: projects/ngx-datatable/src/lib/themes/material.scss [hunk 5]
```diff
@@ -66,6 +66,11 @@ $datatable-pager-active-background: rgba(158, 158, 158, 0.2) !default;
 $datatable-summary-row-background: #ddd !default;
 $datatable-summary-row-background-hover: #ddd !default;
 
+$datatable-ghost-loader-background-light:  #f0f0f0;
+$datatable-ghost-loader-background:  #d8d6d6;
+$datatable-ghost-loader-background-dark:  #ccc7c7;
+
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/themes/material.scss:72]: - Minor spacing adjustment.
- Where did you derive those colors from? Might make sense to sty internally consistent and re-use already existing colors, so the adoption to another design-system is easier.

```suggestion
$datatable-ghost-loader-background-light:  #f5f5f5;
$datatable-ghost-loader-background:  #d9d8d9;
$datatable-ghost-loader-background-dark:  #ccc7c7;
```

File: projects/ngx-datatable/src/lib/components/datatable.component.ts [hunk 6]
```diff
@@ -265,6 +265,12 @@ export class DatatableComponent implements OnInit, DoCheck, AfterViewInit, After
    */
   @Input() loadingIndicator = false;
 
+  /**
+   * Show ghosts loader on each cell.
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/components/datatable.component.ts:269]: Typo I guess -> the `ghost loader` as a whole is plural.

```suggestion
   * Show ghost loaders on each cell.
```

File: projects/ngx-datatable/src/lib/components/body/ghost-loader/ghost-loader.component.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+@Component({
+  selector: `ghost-loader`,
+  templateUrl: `./ghost-loader.component.html`,
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class DataTableGhostLoaderComponent {
+
+  @Input() columns;
+  @Input() pageSize;
+
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/components/body/ghost-loader/ghost-loader.component.ts:12]: Minor spacing alignment to keep it in-sync with the rest of the lib (e.g. https://github.com/siemens/ngx-datatable/blob/master/projects/ngx-datatable/src/lib/components/columns/column-cell.directive.ts):

```suggestion
  @Input() columns;
  @Input() pageSize;
```

**Review by fh1ch — CHANGES_REQUESTED** [review]
@chintankavathia outstanding work, thanks a ton 🙇 I did a lot of nit-picking just so simplify possible back-merges with the origin library, but the result is really nice. Thanks a lot for that 🚀 

Back to you 🏓

**Comment by chintankavathia** [comment]
Thanks @fh1ch for your valuable feedback. I have addressed all your comments. Please have a look again.

**Comment by timowolf** [comment]
Many thanks again. I found an issue in the virtual server-side paging. When pulling down the scroll bar the ghosts are visible and look good. But the scroll bar seems to be blocked. I don't know why this happen yet. But it indicates that something is wrong. During endless scrolling, the ghosts should be visible and the scrollbar should not be blocked.

**Comment by timowolf** [comment]
I also found another topic that I would expect to be different. The current implementation supports paging in the classical way. With that I mean, we see the ghost in a page until the really data is loaded. But with virtual scrolling, I want to see the ghost cells to scroll up and down. It should feel as if the complete table is there, but the ghost row is a placeholder until the real data arrives.

**Comment by timowolf** [comment]
@chintankavathia Many thanks for the great MR. I found two topics concerning the behaviour.

**Comment by chintankavathia** [comment]
> Many thanks again. I found an issue in the virtual server-side paging. When pulling down the scroll bar the ghosts are visible and look good. But the scroll bar seems to be blocked. I don't know why this happen yet. But it indicates that something is wrong. During endless scrolling, the ghosts should be visible and the scrollbar should not be blocked.

Thanks @timowolf for finding that out. I have fixed that issue. it was being caused due to background css used for ghost loader effect.

**Comment by chintankavathia** [comment]
> I also found another topic that I would expect to be different. The current implementation supports paging in the classical way. With that I mean, we see the ghost in a page until the really data is loaded. But with virtual scrolling, I want to see the ghost cells to scroll up and down. It should feel as if the complete table is there, but the ghost row is a placeholder until the real data arrives.

@timowolf I am unable to understand mentioned behaviour. Could you please explain it with more details?

File: projects/ngx-datatable/src/lib/components/body/body.component.ts [hunk 8]
```diff
@@ -544,7 +563,6 @@ export class DataTableBodyComponent implements OnInit, OnDestroy {
     const rowHeight = this.rowDetail.rowHeight;
     return typeof rowHeight === 'function' ? rowHeight(row, index) : (rowHeight as number);
   };
-
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/components/body/body.component.ts:547]: Unrelated change.

File: projects/ngx-datatable/src/lib/themes/material.scss [hunk 9]
```diff
@@ -332,7 +336,6 @@ $datatable-summary-row-background-hover: #ddd !default;
         }
       }
     }
-
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/themes/material.scss:335]: Unrelated change.

File: projects/ngx-datatable/src/lib/themes/material.scss [hunk 10]
```diff
@@ -66,6 +66,10 @@ $datatable-pager-active-background: rgba(158, 158, 158, 0.2) !default;
 $datatable-summary-row-background: #ddd !default;
 $datatable-summary-row-background-hover: #ddd !default;
 
+$datatable-ghost-loader-background-light:  #f5f5f5;
+$datatable-ghost-loader-background:  #d9d8d9;
+$datatable-ghost-loader-background-dark:  #ccc7c7;
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/themes/material.scss:71]: For consistency's sake, we should also default the values here:

```suggestion
$datatable-ghost-loader-background-light: #f5f5f5 !default;
$datatable-ghost-loader-background: #d9d8d9 !default;
$datatable-ghost-loader-background-dark: #ccc7c7 !default;
```

File: projects/ngx-datatable/src/lib/components/body/ghost-loader/ghost-loader.component.html [hunk 11]
```diff
@@ -0,0 +1,13 @@
+<div [style.height]="ghostBodyHeight + 'px'" class="ghost-loader">
+  <div [style.height]="rowHeight + 'px'" class="ghost-element" *ngFor="let item of [].constructor(pageSize)">
+    <ng-container *ngFor="let col of columns">
+      <div class="line"  *ngIf="!col.ghostCellTemplate else ghostCellTemplate;" [style.width]="col?.width + 'px'" >
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/components/body/ghost-loader/ghost-loader.component.html:4]:
```suggestion
      <div class="line" *ngIf="!col.ghostCellTemplate else ghostCellTemplate;" [style.width]="col?.width + 'px'" >
```

File: projects/ngx-datatable/src/lib/themes/bootstrap.scss [hunk 12]
```diff
@@ -111,3 +111,41 @@ bootstrap table theme
     }
   }
 }
+/**
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/themes/bootstrap.scss:114]: I'm not too keen with ngx-datatable's styling hierarchy, but I see a lot of redundant styles for bootstrap, dark and material's CSS files. Is there a way to share utilities and base definitions and e.g. only define/overwrite the colors within the individual themes?

File: projects/ngx-datatable/src/lib/themes/dark.scss [hunk 13]
```diff
@@ -130,3 +130,41 @@
     }
   }
 }
+/**
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/themes/dark.scss:133]: Same here.

**Review by fh1ch — CHANGES_REQUESTED** [review]
@chintankavathia really nice work, also here. Thanks a lot for that 🙇 

The result already looks really good, I've added some code-style findings here.

File: projects/ngx-datatable/src/lib/themes/bootstrap.scss (hunk 12 above)
Comment (chintankavathia) [reply projects/ngx-datatable/src/lib/themes/bootstrap.scss:114]: We in SiMPL-element only using bootstrap.scss. We might need some refactoring and will need to make sure that it doesn't break any style hierarchy. We should take it as separate MR rather than in this. WDYT ?

Comment (fh1ch) [reply projects/ngx-datatable/src/lib/themes/bootstrap.scss:114]: Looping @timowolf and @kfenner here. Doing improvements in a follow-up is fine, but I personally think we should fix the redundant code already as part of this MR. Either by moving the shared SCSS code to the component styling and solely do the color overwrites on theme level or by reducing styling completely and do it on Element level. Not sure which approach is better though.

Comment (chintankavathia) [reply projects/ngx-datatable/src/lib/themes/bootstrap.scss]: I have moved ghost loader specific styles into its own component so no changes in existing ngx-datatable scss.
using variables so we can just override them while using it with SiMPL element

File: projects/ngx-datatable/src/lib/components/body/ghost-loader/ghost-loader.component.scss [hunk 14]
```diff
@@ -0,0 +1,45 @@
+$datatable-ghost-loader-background-light: #f5f5f5 !default;
+$datatable-ghost-loader-background: #d9d8d9 !default;
+$datatable-ghost-loader-background-dark: #ccc7c7 !default;
+$ngx-datatable-default-background: #fff !default;
+$element-radius-0: 0 !default;
```
Comment (timowolf) [inline projects/ngx-datatable/src/lib/components/body/ghost-loader/ghost-loader.component.scss:5]: Unfortunately, this does not work really. These vars get compiled into the component and are not present with the component shipped. Having the styles here is good, the variable part should go into the global theming. so that we can overwrite the vars when importing the theme.

One option I see is to use classes here for the variable parts and these are defined in the global styling.

**Review by timowolf — CHANGES_REQUESTED** [review]
As discussed in the comment, we need to move the variable part out of the component, maybe with classes.

File: projects/ngx-datatable/src/lib/components/body/body.component.ts [hunk 15]
```diff
@@ -89,6 +98,7 @@ import { translateXY } from '../../utils/translate';
           >
           </datatable-body-row>
           <ng-template #groupedRowsTemplate>
+            <ng-container>
```
Comment (timowolf) [inline projects/ngx-datatable/src/lib/components/body/body.component.ts:101]: Why do we need this ng-container. Does not seem to do anything?

File: projects/ngx-datatable/src/lib/components/body/body.component.ts [hunk 16]
```diff
@@ -59,7 +68,7 @@ import { translateXY } from '../../utils/translate';
           [groupedRows]="groupedRows"
           *ngFor="let group of temp; let i = index; trackBy: rowTrackingFn"
           [innerWidth]="innerWidth"
-          [ngStyle]="getRowsStyles(group)"
+          [ngStyle]="getRowsStyles(group, indexes.first + i )"
```
Comment (timowolf) [inline projects/ngx-datatable/src/lib/components/body/body.component.ts:71]: Here is some magic that does not seem to be obvious and we may need some docs. So we iterate through the rows but have not loaded because data is not available, right? 

Why to we need to call this function at all?

File: projects/ngx-datatable/src/lib/components/body/ghost-loader/ghost-loader.component.scss [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+:host.ghost-overlay {
+  position: sticky;
+  top: 20px;
+
+  .ghost-loader {
+    .line {
+      margin: 0.9rem 1.2rem;
+    }
+  }
+}
```
Comment (timowolf) [inline projects/ngx-datatable/src/lib/components/body/ghost-loader/ghost-loader.component.scss:38]: Add new line here

File: projects/ngx-datatable/src/lib/themes/_ghost.scss [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+.ghost-cell-container {
+  background: $datatable-ghost-cell-container-background;
+}
+
+.ghost-cell-strip {
+  background: $datatable-ghost-cell-strip-background;
+  background-image: $datatable-ghost-cell-strip-background-image;
+  border-radius: $datatable-ghost-cell-strip-radius;
+  animation-duration: $datatble-ghost-cell-animation-duration;
+}
```
Comment (timowolf) [inline projects/ngx-datatable/src/lib/themes/_ghost.scss:16]: Empty line at the end of the file

File: projects/ngx-datatable/src/lib/themes/bootstrap.scss [hunk 19]
```diff
@@ -1,6 +1,11 @@
-/*
-bootstrap table theme
-*/
+$datatable-ghost-cell-container-background: #fff !default;
+
+$datatable-ghost-cell-strip-background: #dee2e5 !default;
+$datatable-ghost-cell-strip-background-image: linear-gradient(to right, $datatable-ghost-cell-strip-background 0%, $datatable-ghost-cell-strip-background 10%, #fff, transparent) !default;
+$datatable-ghost-cell-strip-radius: 0 !default;
```
Comment (timowolf) [inline projects/ngx-datatable/src/lib/themes/bootstrap.scss:5]: Why do you use hardcoded variables instead of the bootstrap variables?

**Review by timowolf — CHANGES_REQUESTED** [review]
Awesome work, I just added some review comments and why don't we use the bootstrap variables?

File: projects/ngx-datatable/src/lib/components/body/body.component.ts (hunk 15 above)
Comment (chintankavathia) [reply projects/ngx-datatable/src/lib/components/body/body.component.ts]: Yes. It's not needed. Removed

File: projects/ngx-datatable/src/lib/themes/bootstrap.scss (hunk 19 above)
Comment (chintankavathia) [reply projects/ngx-datatable/src/lib/themes/bootstrap.scss:5]: As discussed, bootstrap variables are not defined in ngx-datatable library hence it throws error we use them here.

File: projects/ngx-datatable/src/lib/components/body/body.component.ts (hunk 16 above)
Comment (chintankavathia) [reply projects/ngx-datatable/src/lib/components/body/body.component.ts:71]: It is basically used for get correct position for ghost cell rows. I added detail comments on `getRowsStyles ` function

Comment (chintankavathia) [reply projects/ngx-datatable/src/lib/components/body/body.component.ts:71]: `ngx-datatable` uses css translate to position each row in correct order when doing scroll. It takes the actual row and finds its position based on that. since in ghost rows we actually don't have real rows we pass just the index to get their position

**Review by timowolf — APPROVED** [review]
Works now fine with me. Tested locally in product. Many thanks.

File: projects/ngx-datatable/src/lib/themes/bootstrap.scss (hunk 12 above)
Comment (timowolf) [reply projects/ngx-datatable/src/lib/themes/bootstrap.scss]: This is done now.
